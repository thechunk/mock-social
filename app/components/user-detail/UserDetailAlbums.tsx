import * as React from 'react';
import {FC, useState} from 'react';
import {View, Text, TouchableHighlight, GestureResponderEvent, Dimensions} from 'react-native';
import AlbumContainer from '../common/album-container/AlbumContainer';
import styles from './styles';
import TouchableImage from "../common/image/TouchableImage";
import {NavigationScreenProps, withNavigation} from "react-navigation";
import * as g from '../../styles/global';

const {width} = Dimensions.get('window');

const UserDetailAlbums: FC<IUserDetailAlbums & NavigationScreenProps> =
    (props: IUserDetailAlbums & NavigationScreenProps) => {
        const limit = 4;
        const imageWidth = Math.floor(width / limit)
            - g.Dimensions.ElementsHorizontalPadding
            - g.Dimensions.Offset / limit;

        const [albums, setAlbums] = useState<Array<IAlbum>>([]);
        const withAlbums = (albums: Array<IAlbum>) => {
            setAlbums(albums);
        };

        const onTapAlbum = (id: number, album: IAlbum) => (event: GestureResponderEvent) => {
            props.navigation.navigate('PhotoGrid', {id, hydrate: album});
        };
        const onTapMore = (id: number) => (event: GestureResponderEvent) => {
            props.navigation.navigate('AlbumList', {id});
        };

        return (
            <View style={[styles.userAlbumsContainer, {height: imageWidth}]}>
                <AlbumContainer
                    userId={props.userId}
                    limit={limit}
                    withAlbums={withAlbums}
                    renderItem={(photo: IPhoto, album: IAlbum)=> (
                        <TouchableImage
                            key={photo.id}
                            style={[g.default.roundedBorder, {width: imageWidth, height: imageWidth}]}
                            onPress={onTapAlbum(photo.albumId, album)}
                            source={{uri: photo.url}}/>
                    )} />
                {albums.length > limit
                    ? (
                        <TouchableHighlight
                            underlayColor={g.Color.White}

                            style={[styles.userAlbumsMore, g.default.roundedBorder, {width: imageWidth, height: imageWidth}]}
                            onPress={onTapMore(props.userId)}>
                            <>
                                <Text style={styles.userAlbumsMoreCount}>{albums.length - limit + 1}</Text>
                                <Text style={styles.userAlbumsMoreText}>more&hellip;</Text>
                            </>
                        </TouchableHighlight>
                    )
                    : null}
            </View>
        )
    }
export default withNavigation(UserDetailAlbums);