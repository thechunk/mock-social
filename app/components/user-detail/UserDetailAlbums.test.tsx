import * as React from 'react';
import {UserDetailAlbums} from './UserDetailAlbums';

import {shallow} from 'enzyme';
import {TouchableHighlight} from "react-native";

describe('UserDetailAlbums', () => {
    const photos = {userId: 1, id: 1, title: 'title', url: 'https://url', thumbnailUrl: 'https://url', albumId: 1};
    const albums = [{userId: 1, id: 1, title: 'title', photos}];
    const navigation = {navigate: jest.fn()};

    beforeEach(() => {
        navigation.navigate.mockClear();
    });

    it('renders correctly', () => {
        expect(shallow(<UserDetailAlbums userId={1} navigation={navigation} initialState={albums} />))
            .toMatchSnapshot();
    });

    it('shows more call to action if more albums than 4 albums', () => {
        const moreAlbums = [1, 2, 3, 4, 5].map(x => ({userId: 1, id: x, title: 'title', photos}));
        const wrapper = shallow(<UserDetailAlbums userId={1} navigation={navigation} initialState={moreAlbums} />);
        expect(wrapper.find(TouchableHighlight)).toHaveLength(1);
    });

    it('does not show call to action if less than 5 albums', () => {
        const wrapper = shallow(<UserDetailAlbums userId={1} navigation={navigation} initialState={albums} />);
        expect(wrapper.find(TouchableHighlight)).toHaveLength(0);
    });

    it('navigates to PhotoGrid on album tap', () => {
        UserDetailAlbums.prototype.onTapAlbum({navigation}, 1, albums[0])();
        expect(navigation.navigate).toBeCalledWith('PhotoGrid', {id: 1, hydrate: albums[0]});
    });

    it('navigates to AlbumList on more tap', () => {
        UserDetailAlbums.prototype.onTapMore({navigation}, 1)();
        expect(navigation.navigate).toBeCalledWith('AlbumList', {id: 1});
    });
});
