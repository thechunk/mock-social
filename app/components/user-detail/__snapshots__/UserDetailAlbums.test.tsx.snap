// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetailAlbums renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
        "marginLeft": 12,
        "marginRight": 12,
      },
      Object {
        "height": 176,
      },
    ]
  }
>
  <AlbumContainer
    limit={4}
    renderItem={[Function]}
    userId={1}
    withAlbums={[Function]}
  />
</View>
`;
